packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -74,35 +75,47 @@ library LibProposing {
             revert L1_TOO_MANY_BLOCKS();
         }
 
-        bytes32 parentMetaHash =
-            _state.blocks[(b.numBlocks - 1) % _config.blockRingBufferSize].metaHash;
-        // assert(parentMetaHash != 0);
+        TaikoData.Block storage parentBlock =
+            _state.blocks[(b.numBlocks - 1) % _config.blockRingBufferSize];
 
         // Check if parent block has the right meta hash. This is to allow the proposer to make sure
         // the block builds on the expected latest chain state.
-        if (params.parentMetaHash != 0 && parentMetaHash != params.parentMetaHash) {
+        if (params.parentMetaHash != 0 && parentBlock.metaHash != params.parentMetaHash) {
             revert L1_UNEXPECTED_PARENT();
         }
 
+        // Verify the passed in L1 state block number.
+        // We only allow the L1 block to be 2 epochs old.
+        // The other constraint is that the L1 block number needs to be larger than or equal the one
+        // in the previous block.
+        if (
+            params.l1StateBlockNumber < block.number - 64
+                || params.l1StateBlockNumber >= block.number
+                || params.l1StateBlockNumber < parentBlock.l1StateBlockNumber
+        ) {
+            revert L1_INVALID_L1_STATE_BLOCK();
+        }
+
         // Initialize metadata to compute a metaHash, which forms a part of
         // the block data to be stored on-chain for future integrity checks.
         // If we choose to persist all data fields in the metadata, it will
         // require additional storage slots.
         unchecked {
             meta_ = TaikoData.BlockMetadata({
-                l1Hash: blockhash(block.number - 1),
+                l1Hash: blockhash(params.l1StateBlockNumber),
                 difficulty: 0, // to be initialized below
                 blobHash: 0, // to be initialized below
                 extraData: params.extraData,
                 depositsHash: _EMPTY_ETH_DEPOSIT_HASH,
                 coinbase: params.coinbase,
                 id: b.numBlocks,
                 gasLimit: _config.blockMaxGasLimit,
-                timestamp: uint64(block.timestamp),
-                l1Height: uint64(block.number - 1),
+                // Use the timestamp of the L1 state block
+                timestamp: uint64(block.timestamp - 12 * (block.number - params.l1StateBlockNumber)),
```

Comment:
if `params.l1StateBlockNumber == block.number` is not allowed, L2's timestamp is always at least 12 second smaller than L1's `block.timestamp`. Shall we use `uint64(block.timestamp - 12 * (block.number  - 1 - params.l1StateBlockNumber))` instead?

Reply:
Since L2's block time is not strictly 12 seconds, calculating L2's timestamp using l1StateBlockNumber is always not accurate, maybe we can simply just use `block.timestamp`?

Reply:
> Since L2's block time is not strictly 12 seconds, calculating L2's timestamp using l1StateBlockNumber is always not accurate, maybe we can simply just use `block.timestamp`?
 
L1s block time is always 12 seconds, so I think it's the best way we can calculate the latest timestamp while still making sure that the value is predictable for the proposer who starts building an L2 block using `l1StateBlockNumber `. Using `block.timestamp` would make it so the preconfer HAS to include the L2 block being in some specific L1 block to make sure the L2 block execution goes as expected (which the preconfer cannot always do).